% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/figures.R
\name{plotTaxonomy}
\alias{plotTaxonomy}
\title{Barplot of the most abundant taxa in a SQM object}
\usage{
plotTaxonomy(SQM, rank = "phylum", count = "percent", N = 15,
  tax = NULL, others = T, color = NULL, rescale = F,
  ignore_unclassified = F)
}
\arguments{
\item{SQM}{A SQM object.}

\item{rank}{Taxonomic rank to plot (default \code{phylum}).}

\item{count}{character. Either \code{"percent"} for percentages, or \code{"abund"} for raw abundances (default \code{"percent"}).}

\item{N}{integer Plot the \code{N} most abundant taxa (default \code{15}).}

\item{tax}{character. Custom taxa to plot. If provided, it will override \code{N} (default \code{NULL}).}

\item{others}{logical. Collapse the abundances of least abundant taxa, and include the result in the plot (default \code{TRUE}).}

\item{color}{Vector with custom colors for the different features. If empty, we will use our own hand-picked pallete if N<=15, and the default ggplot2 palette otherwise (default \code{NULL}).}

\item{rescale}{logical. Re-scale results to percentages (default \code{FALSE}).}

\item{ignore_unclassified}{logical. Don't include unclassified contigs in the plot (default \code{FALSE}).}
}
\value{
a ggplot2 plot object.
}
\description{
This function selects the most abundant taxa across all samples in a SQM object and represents their abundances in a barplot. Alternatively, a custom set of taxa can be represented.
}
\examples{
data(Hadza)
Hadza.amin = subsetFun(Hadza, 'Amino acid metabolism')
# Taxonomic distribution of amino acid metabolism ORFs at the family level.
plotTaxonomy(Hadza.amin, 'family')
}
\seealso{
\code{\link[plotFunctions]{plotFunctions}} for plotting the most abundant functions of a SQM object; \code{\link[plotBars]{plotBars}} and \code{\link[plotBars]{plotHeatmap}} for plotting barplots or heatmaps with arbitrary data.
}
